src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectChallengeProperties.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+
+namespace Microsoft.AspNetCore.Authentication.OpenIdConnect
+{
+    public class OpenIdConnectChallengeProperties : OAuthChallengeProperties
+    {
+        /// <summary>
+        /// The property for the MaxAge parameter value being used for a challenge request.
+        /// </summary>
+        public static readonly string MaxAgeParameterKey = OpenIdConnectParameterNames.MaxAge;
+
+        /// <summary>
+        /// The property for the Prompt parameter value being used for a challenge request.
+        /// </summary>
+        public static readonly string PromptParameterKey = OpenIdConnectParameterNames.Prompt;
+
+        public OpenIdConnectChallengeProperties()
+        { }
+
+        public OpenIdConnectChallengeProperties(IDictionary<string, string> items)
+            : base(items)
+        { }
+
+        public OpenIdConnectChallengeProperties(IDictionary<string, string> items, IDictionary<string, object> parameters)
+            : base(items, parameters)
+        { }
+
+        /// <summary>
+        /// The "max_age" parameter value being used for a challenge request.
+        /// </summary>
+        public TimeSpan? MaxAge
+        {
+            get => GetParameter<TimeSpan?>(MaxAgeParameterKey);
+            set => SetParameter(MaxAgeParameterKey, value);
+        }
+
+        /// <summary>
+        /// The "prompt" parameter value being used for a challenge request.
+        /// </summary>
+        public string Prompt
+        {
+            get => GetParameter<string>(PromptParameterKey);
```

Comment:
Its a shame these getters are basically useless

Reply:
Is it too crazy to also have a set of internal static methods for the getters so you can do something like:

```C#
public string Prompt {
   get => ComputePrompt(this);
   set => SetParameter(PromptParameterKey, value);
}

 internal static string ComputePrompt(AuthenticationProperties properties, string defaultValue = null) 
   => properties.GetParameter<string>(OpenIdConnectionParameterNames.Prompt) ?? defaultValue;
```

Then the handlers code would just be `OpenIdChallengeProperties.ComputePrompt(properties, Options.Prompt)` which is a bit nicer since the constant doesn't show up

Reply:
This would also let us keep `GetParameter` protected on `AuthenticationProperties` if the handlers use something else to compute the parameters.

Reply:
I wish [Extension Everything](https://github.com/dotnet/csharplang/issues/192) was already here… that would be so perfect for this…

I’m not entirely convinced that having a set of public static methods spread over multiple types is necessarily cleaner than a single public method that uses constants. Sure, we could make `GetParameter` protected, and make the constants even private, but those static methods do also appear pretty verbose.

And unless something else is actually decided on the other PR, I would keep those methods public on the `AuthenticationProperties` (someone needs to make a final decision there).

Reply:
Why can't the handlers use them?

Reply:
@Tratcher @HaoK was referring to the properties which the handlers cannot use if we want them to operate on the general `AuthenticationProperties` type instead of one of the subtypes. It would be nice if the handlers could just do `properties.Prompt` as well but they have to use `properties.GetParameter<string>("prompt")`.

Reply:
Yeah that's the main reason I proposed the 'real' public static getter that the handlers could use, and then getter would just route to...

Reply:
Shrug, I don't think an API wrapper around GetParameter is needed if the handler is the only caller.

Reply:
The main reason to add these would only be if we did not want GetParameters to show up as a public API on each derived ChallengeProperties class. I can live without this, I just think this is slightly cleaner, and if we wanted to keep the helpers hidden like we had before with Items.

Reply:
That’s what I was referring t0 in my [summary below](https://github.com/aspnet/Security/pull/1691#issuecomment-375145244) (or is it above?!). With the current state of the other PR and David’s remark, I’m not sure if we want to keep the API public now or rather not. If we end up making them protected, then having such static helpers would probably be a good idea to keep the logic out of the handlers.

Reply:
So I'm pretty happy overall with how this looks now, and I think the Parameters addition is a huge win, unless @davidfowl feels very strongly that we should keep the `properties.GetParameter<T>`method hidden.  Personally if I was implementing this, I would add the additional public statics, but its not a big deal, so I'm happy to defer to you as the PR author in this case, to make the call on whether its worth adding the additional public statics...

Reply:
No more creep... we need to get this wrapped up.

Reply:
So this is only 'creep' if `GetParameters` stays public.  If this ends up as protected in the other PR, this becomse mandatory...